app/models/user.rb

```diff
@@ -25,80 +29,165 @@ def to_s
     webauth.attributes['DISPLAYNAME'] || webauth.login
   end
 
-  @role_cache = {}
-
   # Queries Solr for a given record, returning synthesized role strings that are defined. Searches:
   # (1) for User by sunet id
   # (2) groups actually contains the sunetid, so it is just looking at different solr fields
   # NOTE: includes legacy roles that have to be translated
-  # @param [String] DRUID, fully qualified
-  # @return [Array[String]] list of roles
+  # @param [String] DRUID
+  # @return [Array<String>] set of roles permitted by DRUID
   def roles(pid)
+    druid = DruidTools::Druid.new(pid).druid
     @role_cache ||= {}
-    return @role_cache[pid] if @role_cache[pid]
-
-    resp = Dor::SearchService.query('id:"' + pid + '"')['response']['docs'].first || {}
-    toret = []
-    my_groups = groups
-    %w(apo_role_group_manager_ssim apo_role_person_manager_ssim).each do |key|
-      toret << 'dor-apo-manager' if resp[key] && (resp[key] & my_groups).length > 0
-    end
-
+    return @role_cache[druid] if @role_cache[druid]
+    # Try to retrieve a Solr doc for the DRUID
+    obj_doc = Dor::SearchService.query('id:"' + druid + '"')['response']['docs'].first || {}
+    return [] if obj_doc.empty?
+    druid_roles = Set.new
+    # Determine whether user has 'dor-apo-manager' role
+    apo_roles = %w(apo_role_group_manager_ssim apo_role_person_manager_ssim)
```

Comment:
Next commit will use a `solr` prefix for these, e.g. `solr_apo_roles` and `solr_role_allowed?` (to address comment below).